torch/csrc/Device.cpp

```diff
@@ -1,235 +1,106 @@
-    return THPDevice_New(device);
-  }
-  Py_RETURN_NONE;
-  END_HANDLE_TH_ERRORS
+inline Device parseDeviceIndex(DeviceType type, DeviceIndex index) {
+  // -1 is allowed in ATen/C++, to mean the default device, but not in Python
+  TORCH_CHECK(index >= 0, "Device index must not be negative");
+  return Device(type, index);
 }
 
-PyObject *THPDevice_type(THPDevice *self, PyObject *noargs)
-{
-  HANDLE_TH_ERRORS
+inline std::string deviceType(const Device& device) {
   std::ostringstream oss;
-  oss << self->device.type();
-  return THPUtils_packString(oss.str().c_str());
-  Py_RETURN_NONE;
-  END_HANDLE_TH_ERRORS
-}
-
-PyObject *THPDevice_index(THPDevice *self, PyObject *noargs)
-{
-  HANDLE_TH_ERRORS
-  if (self->device.has_index()) {
-    return THPUtils_packInt64(self->device.index());
-  } else {
-    Py_RETURN_NONE;
-  }
-  END_HANDLE_TH_ERRORS
-}
-
-static Py_ssize_t THPDevice_hash(THPDevice *self)
-{
-  HANDLE_TH_ERRORS
-  return static_cast<Py_ssize_t>(std::hash<at::Device>{}(self->device) % std::numeric_limits<Py_ssize_t>::max());
-  END_HANDLE_TH_ERRORS_RET(-1)
+  oss << device.type();
+  return oss.str();
 }
 
-PyObject *THPDevice_rc(PyObject *a, PyObject *b, int op) {
-  HANDLE_TH_ERRORS
-  if (!THPDevice_Check(a) || !THPDevice_Check(b)) {
-    // Py_RETURN_NOTIMPLEMENTED not in python 2.
-    Py_INCREF(Py_NotImplemented);
-    return Py_NotImplemented;
-  }
-  THPDevice *da = reinterpret_cast<THPDevice*>(a);
-  THPDevice *db = reinterpret_cast<THPDevice*>(b);
-
-  switch(op) {
-    case Py_EQ:
-      if (da->device == db->device) {
-        Py_RETURN_TRUE;
-      } else {
-        Py_RETURN_FALSE;
-      }
-    case Py_NE:
-      if (da->device == db->device) {
-        Py_RETURN_FALSE;
-      } else {
-        Py_RETURN_TRUE;
-      }
-    case Py_LT:
-    case Py_LE:
-    case Py_GT:
-    case Py_GE:
-      throw torch::TypeError("comparison not implemented");
-    default:
-      throw torch::TypeError("unexpected comparison op");
-  }
-  END_HANDLE_TH_ERRORS
-}
+} // namespace
 
-PyObject *THPDevice_reduce(PyObject *_self, PyObject *noargs)
-{
-  HANDLE_TH_ERRORS
-  auto self = (THPDevice*)_self;
-  auto ret = THPObjectPtr{PyTuple_New(2)};
-  if (!ret) throw python_error();
-
-  py::object torch_module = py::module::import("torch");
-  py::object torch_device = torch_module.attr("device");
-  PyTuple_SET_ITEM(ret.get(), 0, torch_device.release().ptr());
-
-  THPObjectPtr args;
-  std::ostringstream oss;
-  oss << self->device.type();
-  if (self->device.has_index()) {
-    args = THPObjectPtr{Py_BuildValue("(si)", oss.str().c_str(), self->device.index())};
+Device parseDevice(py::handle device) {
+  if (py::isinstance<Device>(device)) {
+    return device.cast<Device>();
+  } else if (py::isinstance<py::str>(device)) {
+    return Device(device.cast<std::string>());
+  } else if (py::isinstance<py::int_>(device)) {
+    return parseDeviceIndex(kCUDA, device.cast<DeviceIndex>());
```

Comment:
hmm.. Isn't this adding a new constructor- `torch.device(2)` creates a cuda device? I don't see that constructor in the old version of the code

Reply:
`parseDevice` is just replicating the exact behavior of the old `THPDevice_pynew` function here.